server/sonar-server/src/test/java/org/sonar/server/measure/ws/SearchHistoryActionTest.java

```diff
@@ -0,0 +1,386 @@
+      .map(a -> formatDateTime(a.getCreatedAt()))
+      .collect(Collectors.toList());
+    db.commit();
+    wsRequest.setComponent(project.getKey()).setFrom(analysisDates.get(1)).setTo(analysisDates.get(3));
+
+    SearchHistoryResponse result = call();
+
+    assertThat(result.getPaging()).extracting(Paging::getPageIndex, Paging::getPageSize, Paging::getTotal).containsExactly(1, 100, 3);
+    assertThat(result.getMeasures(0).getHistoryList()).extracting(HistoryValue::getDate).containsExactly(
+      analysisDates.get(1), analysisDates.get(2), analysisDates.get(3));
+  }
+
+  @Test
+  public void return_best_values_for_files() {
+    dbClient.metricDao().insert(dbSession, newMetricDto().setKey("optimized").setValueType(ValueType.INT.name()).setOptimizedBestValue(true).setBestValue(456d));
+    dbClient.metricDao().insert(dbSession, newMetricDto().setKey("new_optimized").setValueType(ValueType.INT.name()).setOptimizedBestValue(true).setBestValue(789d));
+    db.commit();
+    ComponentDto file = db.components().insertComponent(newFileDto(project));
+    wsRequest.setComponent(file.getKey()).setMetrics(Arrays.asList("optimized", "new_optimized"));
+
+    SearchHistoryResponse result = call();
+
+    assertThat(result.getMeasuresCount()).isEqualTo(2);
+    assertThat(result.getMeasuresList().get(0).getHistoryList()).extracting(HistoryValue::getValue).containsExactly("789");
+    assertThat(result.getMeasuresList().get(1).getHistoryList()).extracting(HistoryValue::getValue).containsExactly("456");
+
+    // Best value is not applied to project
+    wsRequest.setComponent(project.getKey());
+    result = call();
+    assertThat(result.getMeasuresList().get(0).getHistoryCount()).isEqualTo(0);
+  }
+
+  @Test
+  public void do_not_return_unprocessed_analyses() {
+    dbClient.snapshotDao().insert(dbSession, newAnalysis(project).setStatus(STATUS_UNPROCESSED));
+    db.commit();
+
+    SearchHistoryResponse result = call();
+
+    // one analysis in setUp method
+    assertThat(result.getPaging().getTotal()).isEqualTo(1);
+  }
+
+  @Test
+  public void do_not_return_developer_measures() {
+    wsRequest.setMetrics(singletonList(complexityMetric.getKey()));
+    dbClient.measureDao().insert(dbSession, newMeasureDto(complexityMetric, project, analysis).setDeveloperId(42L));
+    db.commit();
+
+    SearchHistoryResponse result = call();
+
+    assertThat(result.getMeasuresCount()).isEqualTo(1);
+    assertThat(result.getMeasures(0).getHistoryCount()).isEqualTo(0);
+  }
+
+  @Test
+  public void fail_if_unknown_metric() {
+    wsRequest.setMetrics(newArrayList(complexityMetric.getKey(), nclocMetric.getKey(), "METRIC_42", "42_METRIC"));
+
+    expectedException.expect(IllegalArgumentException.class);
+    expectedException.expectMessage("Metrics 42_METRIC, METRIC_42 are not found");
+
+    call();
+  }
+
+  @Test
+  public void fail_if_not_enough_permissions() {
```

Comment:
I don't understand the content of this test, from where is coming the ForbiddenException ?
Anonymous user should have access to the project if the "Browse" permission is set to the project.
It's not clear for me what is really tested here, reading the test I'm thinking that this WS requires to be authenticated which is not the case.